wdl-analysis/src/analyzer.rs

```diff
@@ -0,0 +1,478 @@
+//! Implementation of the analyzer.
+
+use std::fmt;
+use std::future::Future;
+use std::mem::ManuallyDrop;
+use std::ops::Range;
+use std::path::absolute;
+use std::path::Path;
+use std::path::PathBuf;
+use std::sync::Arc;
+use std::thread::JoinHandle;
+
+use line_index::LineCol;
+use line_index::LineIndex;
+use line_index::WideEncoding;
+use line_index::WideLineCol;
+use path_clean::clean;
+use rowan::GreenNode;
+use tokio::runtime::Handle;
+use tokio::sync::mpsc;
+use tokio::sync::oneshot;
+use url::Url;
+use wdl_ast::AstNode;
+use wdl_ast::Diagnostic;
+use wdl_ast::SyntaxNode;
+use wdl_ast::Validator;
+
+use crate::graph::DocumentGraphNode;
+use crate::graph::ParseState;
+use crate::queue::AnalysisQueue;
+use crate::queue::AnalyzeRequest;
+use crate::queue::RemoveRequest;
+use crate::queue::Request;
+use crate::rayon::RayonHandle;
+use crate::DocumentScope;
+
+/// Represents the kind of analysis progress being reported.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum ProgressKind {
+    /// The progress is for parsing documents.
+    Parsing,
+    /// The progress is for analyzing documents.
+    Analyzing,
+}
+
+impl fmt::Display for ProgressKind {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Self::Parsing => write!(f, "parsing"),
+            Self::Analyzing => write!(f, "analyzing"),
+        }
+    }
+}
+
+/// Converts a local path to a file schemed URI.
+pub fn path_to_uri(path: &Path) -> Option<Url> {
+    Url::from_file_path(clean(absolute(path).ok()?)).ok()
+}
+
+/// Represents the result of an analysis.
+///
+/// Analysis results are cheap to clone.
+#[derive(Debug, Clone)]
+pub struct AnalysisResult {
+    /// The URI of the analyzed document.
+    uri: Arc<Url>,
+    /// The root node of the document.
+    ///
+    /// This is `None` if the document failed to be read.
+    root: Option<GreenNode>,
+    /// The line index used to map line/column offsets to byte offsets and vice
+    /// versa.
+    lines: Option<Arc<LineIndex>>,
```

Comment:
When is this `None`?

Reply:
It's `None` when `root` is also `None`, indicating that there was an error fetching the document (i.e. doesn't exist on disk or HTTP error).

These three fields (`root`, `lines`, and `error`) are probably better encapsulated in an enum. I'll write that down for a follow-up fix.

Reply:
I pushed up a fix for this.